### PR TITLE
Data review / MongoDB / Docker

### PR DESCRIPTION
**Pull request by jfavlam**
## Quickstart local setup

- `git clone git@github.com:MediaComem/open-co2.git`
- `cd open-co2/server`
- `docker-compose up -d`

Default GQL endpoint is available at http://localhost:4000/graphql

Query example using Apollo Studio (getCategoryByName):
```
query getCategoryByName($name: String!) {
  getCategoryByName(name: $name) {
    title
    fullPath
  }
}
```
Variable:
```
{
  "name": "freight_train"
}
```

---

## Main features/fix

- Setup XLS source file (https://docs.google.com/spreadsheets/d/168Gjlb9d5b_0Vb4q1IeKWGPGBIEcD6oo0ggLZafygwY)
- Seeder to process XLS file
- Persist data in DB (MongoDB)
- Review GraphQL types definition
- Start resolver refactor to get data from DB (all units, all categories, category by name)
- Review project structure (/server/app + /server/seeder)
- Dockerizing project
  - Seed data
  - Populate MongoDB
- GraphQL Playground service in stack (as Apollo alternative) - At http://localhost:4200 by default
- Open CO2 server landing page – At http://localhost:4000 by default

Notes/Missing:
- Resolver and available API queries are incomplete at this stage
- co2eq model will be review to calculate confidence index in mean values + categories with different children units will be ignored

### CODE REVIEW

File: server/seeder/modules/fileComposer/fileComposer.js [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+      // Loop over level's names
+      for (let n = 0; n < this.maxDepth; n++) {
+        if (row[`Level ${n + 1}`] !== undefined) {
+          // Add title key with value from level
+          row.title = row[`Level ${n + 1}`];
+          // Add depth key from level number
+          const currentLevel = Object.keys(row)[0].replace("Level ", "");
+          row.depth = parseInt(currentLevel);
+          // Delete unused level key
+          delete row[`Level ${n + 1}`];
```
Comment (slecorne) [inline server/seeder/modules/fileComposer/fileComposer.js]: Not sure but I would say the level is n+1
```suggestion
        const currentLevel = n+1;
        const key = `Level ${currentLevel }`;
        if (row[key] !== undefined) {
          // Add title key with value from level
          row.title = row[key];
          // Add depth key from level number
          row.depth = parseInt(currentLevel);
          // Delete unused level key
          delete row[key];
          break;
```

File: server/seeder/modules/fileComposer/fileComposer.js [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+      delete row[`Description`];
+    }
+  }
+
+  /**
+   * Group internal methods and execute in order to process categories sheet
+   */
+  processCategories() {
+    // 1 – Update object keys from sheet headers
+    this.#updateKeys();
```
Comment (slecorne) [inline server/seeder/modules/fileComposer/fileComposer.js]:
```suggestion
    this.#parseRows();
```

File: server/seeder/modules/fileComposer/fileComposer.js [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    // Loop over rows
+    for (let i = 0, l = this.sheet.length; i < l; i++) {
+      const row = this.sheet[i];
+      // Generate hash from fullpath
+      row.categoryId = hashString(row.fullPath);
+    }
+  }
+
+  #addChildrens() {
+    const deepTree = this.#getDeepTree();
```
Comment (slecorne) [inline server/seeder/modules/fileComposer/fileComposer.js]: avoid parsing the tree multiple times. Do once after ParseRows and store in private var

File: server/seeder/modules/fileComposer/fileComposer.js [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+/**
+ * Class to process XLS sheets and turn it to JS objets
+ */
+export default class FileComposer {
+  /**
+   * FileComposer constructor
+   * @param {Sheet content from fileReader} sheet
+   */
+  constructor(sheet) {
+    this.sheet = sheet;
```
Comment (slecorne) [inline server/seeder/modules/fileComposer/fileComposer.js]:
```suggestion
   * @param {Sheet content from fileReader} rows
   */
  constructor(rows) {
    this.rows= rows;
```

File: server/seeder/modules/fileReader/fileReader.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import XLSX from "xlsx";
+
+/**
+ * Class to read XLS file and extract sheets content
+ */
+export default class FileReader {
```
Comment (slecorne) [inline server/seeder/modules/fileReader/fileReader.js]:
```suggestion
export default class XLSDataReader{
```

File: server/seeder/modules/fileComposer/fileComposer.js [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+  hashString,
+  getMeanFromArray,
+  getDeviationFromArray
+} from "../../helpers/utils.js";
+// import util from "util";
+
+/**
+ * Class to process XLS sheets and turn it to JS objets
+ */
+export default class FileComposer {
```
Comment (slecorne) [inline server/seeder/modules/fileComposer/fileComposer.js]:
```suggestion
export default class DataParser{
```

File: server/seeder/modules/fileComposer/fileComposer.js [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+      for (const child of tree.children) {
+        const tmp = this.#findInTree(name, child);
+        if (!(Object.keys(tmp).length === 0 && tmp.constructor === Object)) {
+          tmp.path.unshift(tree.name);
+          return tmp;
+        }
+      }
+      return {};
+    }
+  }
```
Comment (slecorne) [inline server/seeder/modules/fileComposer/fileComposer.js]: really hard to follow, I will check if I can help to make it more readable

**Review by slecorne — CHANGES_REQUESTED** [review]
I put quite a few comments on naming, we can discuss if it really make sense or not

File: server/seeder/modules/fileComposer/fileComposer.js (hunk 3 above)
Comment (jfavlam) [reply server/seeder/modules/fileComposer/fileComposer.js]: Very good point, especially with this function 👍

File: server/seeder/modules/fileComposer/fileComposer.js (hunk 1 above)
Comment (jfavlam) [reply server/seeder/modules/fileComposer/fileComposer.js]: row[key] https://usbeketrica.com/media/83950/download/rocky2.jpg?v=1&inline=1
